File: src/main/java/de/swt/testing/ShoppingCart.java
package de.swt.testing;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * A shopping cart that holds items with prices.
 */
public class ShoppingCart {

    private final List<Item> items = new ArrayList<>();

    /**
     * Adds an item to the cart.
     *
     * @param name the item name
     * @param price the item price
     */
    public void addItem(String name, double price) {
        items.add(new Item(name, price));
    }

    /**
     * Returns the total price of all items in the cart.
     */
    public double getTotal() {
        double total = 0;
        for (Item item : items) {
            total += item.price();
        }
        return total;
    }

    /**
     * Returns the number of items in the cart.
     */
    public int getItemCount() {
        return items.size();
    }

    /**
     * Returns an unmodifiable view of the items in the cart.
     */
    public List<Item> getItems() {
        return Collections.unmodifiableList(items);
    }

    /**
     * Removes all items from the cart.
     */
    public void clear() {
        items.clear();
    }

    /**
     * An item in the shopping cart.
     */
    public record Item(String name, double price) {}
}

File: python/src/shopping_cart.py
"""A shopping cart that holds items with prices."""

from dataclasses import dataclass


@dataclass(frozen=True)
class Item:
    """An item in the shopping cart."""

    name: str
    price: float


class ShoppingCart:
    """A shopping cart that holds items with prices."""

    def __init__(self):
        self._items: list[Item] = []

    def add_item(self, name, price):
        """Add an item to the cart."""
        self._items.append(Item(name, price))

    def get_total(self):
        """Return the total price of all items in the cart."""
        return sum(item.price for item in self._items)

    def get_item_count(self):
        """Return the number of items in the cart."""
        return len(self._items)

    def get_items(self):
        """Return a copy of the items list."""
        return list(self._items)

    def clear(self):
        """Remove all items from the cart."""
        self._items.clear()
